### PR TITLE
Lnet state

### PR DESCRIPTION
**Pull request by johnsonw**
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2438)

### CODE REVIEW

**Comment by johnsonw** [comment]
When the iml-network service is started it will initialize two caches:
1. interface cache
2. lnet cache

The cache will be use to ensure that database transactions are only executed when data that is received from the agents changes. Here is some data when the service is started with some existing data:

```
interface cache: {4: [NetworkInterface { interface: "eth1", mac_address: Some("08:00:27:06:77:0c"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.20.22, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe06:770c, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth2", mac_address: Some("08:00:27:38:97:98"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.10.22, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe38:9798, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth3", mac_address: Some("08:00:27:a8:8a:77"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.40.22, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fea8:8a77, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth4", mac_address: Some("08:00:27:0c:11:7f"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.50.22, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe0c:117f, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth5", mac_address: Some("08:00:27:73:42:a8"), interface_type: Some(Tcp), inet4_address: [], inet6_address: [Ipv6Network { addr: fe80::5f60:a22f:f2ad:115b, prefix: 64 }], stats: None, is_up: true, is_slave: false }], 1: [NetworkInterface { interface: "eth1", mac_address: Some("08:00:27:0e:1e:b4"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.20.11, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe0e:1eb4, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth2", mac_address: Some("08:00:27:c5:1d:e0"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.10.11, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fec5:1de0, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth3", mac_address: Some("08:00:27:91:7e:9c"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.40.11, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe91:7e9c, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth4", mac_address: Some("08:00:27:de:00:88"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.50.11, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fede:88, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth5", mac_address: Some("08:00:27:ef:24:d7"), interface_type: Some(Tcp), inet4_address: [], inet6_address: [Ipv6Network { addr: fe80::5184:c009:26cd:d225, prefix: 64 }], stats: None, is_up: true, is_slave: false }], 3: [NetworkInterface { interface: "eth1", mac_address: Some("08:00:27:d4:97:94"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.20.21, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fed4:9794, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth2", mac_address: Some("08:00:27:0c:da:21"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.10.21, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe0c:da21, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth3", mac_address: Some("08:00:27:a5:4e:1b"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.40.21, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fea5:4e1b, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth4", mac_address: Some("08:00:27:41:15:b6"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.50.21, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe41:15b6, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth5", mac_address: Some("08:00:27:50:d2:26"), interface_type: Some(Tcp), inet4_address: [], inet6_address: [Ipv6Network { addr: fe80::b914:cb8f:8eb0:97ce, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth0", mac_address: Some("52:54:00:fd:32:21"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.0.2.15, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::5054:ff:fefd:3221, prefix: 64 }], stats: None, is_up: true, is_slave: false }], 5: [NetworkInterface { interface: "eth1", mac_address: Some("08:00:27:69:39:24"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.20.12, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe69:3924, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth2", mac_address: Some("08:00:27:81:32:59"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.10.12, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe81:3259, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth3", mac_address: Some("08:00:27:5e:c3:99"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.40.12, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe5e:c399, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth4", mac_address: Some("08:00:27:52:bf:e9"), interface_type: Some(Tcp), inet4_address: [Ipv4Network { addr: 10.73.50.12, prefix: 24 }], inet6_address: [Ipv6Network { addr: fe80::a00:27ff:fe52:bfe9, prefix: 64 }], stats: None, is_up: true, is_slave: false }, NetworkInterface { interface: "eth5", mac_address: Some("08:00:27:51:0a:8a"), interface_type: Some(Tcp), inet4_address: [], inet6_address: [Ipv6Network { addr: fe80::64:6a8b:161f:6939, prefix: 64 }], stats: None, is_up: true, is_slave: false }]}
Dec 16 19:29:41 adm.local iml-network[27511]: Dec 16 19:29:41.208 DEBUG iml_network: lnet cache: {1: LNet { net: [Net { net_type: "tcp", local_nis: [Nid { nid: "10.73.20.11@tcp", status: "up", interfaces: Some({0: "eth1"}) }] }], state: Up }, 5: LNet { net: [Net { net_type: "tcp", local_nis: [Nid { nid: "10.73.20.12@tcp", status: "up", interfaces: Some({0: "eth1"}) }] }], state: Up }, 4: LNet { net: [Net { net_type: "tcp", local_nis: [Nid { nid: "10.73.20.22@tcp", status: "up", interfaces: Some({0: "eth1"}) }] }], state: Up }, 3: LNet { net: [Net { net_type: "tcp", local_nis: [Nid { nid: "10.73.20.21@tcp", status: "up", interfaces: Some({0: "eth1"}) }] }], state: Up }}
```

Next, I tested the state changes in the `lnet` and `nid` tables, starting from an unconfigured state on mds2.

```
[root@mds2 ~]# lustre_rmmod ptlrpc
[root@mds2 ~]# lnetctl lnet unconfigure

chroma=> select * from nid;
 id | net_type | host_id |       nid       | status | interfaces
----+----------+---------+-----------------+--------+------------
  4 | tcp      |       3 | 10.73.20.21@tcp | up     | {eth1}
  1 | tcp      |       1 | 10.73.20.11@tcp | up     | {eth1}
  3 | tcp      |       4 | 10.73.20.22@tcp | up     | {eth1}
(3 rows)

chroma=> select * from lnet;
  id   | host_id | state | nids
-------+---------+-------+------
     4 |       3 | up    | {4}
     1 |       1 | up    | {1}
     3 |       4 | up    | {3}
 17250 |       5 | down  | {}
```

Notice the id is 17,250. The id initially incremented very quickly because a cache was not implemented initially, which would run an upsert transaction every time data is received from the storage server. The upsert causes the id to increment, regardless of whether or not data changed. Since the cache was implemented, the id is not incrementing unnecessarily. 

Next, I configure lnet on mds2.

```
[root@mds2 ~]# lnetctl lnet configure

// Nids should not change yet
chroma=> select * from nid;
 id | net_type | host_id |       nid       | status | interfaces
----+----------+---------+-----------------+--------+------------
  4 | tcp      |       3 | 10.73.20.21@tcp | up     | {eth1}
  1 | tcp      |       1 | 10.73.20.11@tcp | up     | {eth1}
  3 | tcp      |       4 | 10.73.20.22@tcp | up     | {eth1}
(3 rows)

chroma=> select * from lnet;
  id   | host_id | state | nids
-------+---------+-------+------
     4 |       3 | up    | {4}
     1 |       1 | up    | {1}
     3 |       4 | up    | {3}
 17250 |       5 | up    | {}
(4 rows)
```

Notice the state has changed to `up`. Next, I will add a network:

```
[root@mds2 ~]# lnetctl net add --net tcp0 --if eth1

chroma=> select * from nid;
  id   | net_type | host_id |       nid       | status | interfaces
-------+----------+---------+-----------------+--------+------------
     4 | tcp      |       3 | 10.73.20.21@tcp | up     | {eth1}
     1 | tcp      |       1 | 10.73.20.11@tcp | up     | {eth1}
     3 | tcp      |       4 | 10.73.20.22@tcp | up     | {eth1}
 17289 | tcp      |       5 | 10.73.20.12@tcp | up     | {eth1}

chroma=> select * from lnet;
  id   | host_id | state |  nids
-------+---------+-------+---------
     4 |       3 | up    | {4}
     1 |       1 | up    | {1}
     3 |       4 | up    | {3}
 17250 |       5 | up    | {17289}
(4 rows)
```

The nid was associated with the host as expected. Now, I will unconfigure lnet:

```
[root@mds2 ~]# lustre_rmmod ptlrpc
[root@mds2 ~]# lnetctl lnet unconfigure
[root@mds2 ~]# lustre_rmmod
[root@mds2 ~]# udevadm info --path /sys/module/lnet
syspath not found
[root@mds2 ~]# echo $?
2

chroma=> select * from lnet;
  id   | host_id |  state   | nids
-------+---------+----------+------
     4 |       3 | up       | {4}
     1 |       1 | up       | {1}
     3 |       4 | up       | {3}
 17250 |       5 | unloaded | {}
```
Notice the state is accurately shown as `unloaded`.

File: iml-services/iml-network/src/main.rs [hunk 1]
```diff
@@ -96,6 +101,28 @@ async fn update_interfaces(
...
 }
 
+async fn handle_interfaces(
+    pool: &PgPool,
+    host_id: i32,
+    interfaces: &[NetworkInterface],
+    interface_cache: &mut HashMap<i32, Vec<NetworkInterface>>,
+) -> Result<(), sqlx::Error> {
+    if let Some(cached_interfaces) = interface_cache.get(&host_id) {
+        if cached_interfaces.as_slice() != interfaces {
```
Comment (utopiabound) [inline iml-services/iml-network/src/main.rs:111]: Are these always consistently sorted?

File: iml-services/iml-network/src/main.rs [hunk 2]
```diff
@@ -118,23 +145,28 @@ async fn update_network_stats(
...
     lnet_data: &LNet,
     host_id: i32,
-) -> Option<(Vec<String>, Vec<i32>, Vec<String>, Vec<String>, Vec<String>)> {
-    if lnet_data.net.is_empty() {
-        return None;
+) -> (
+    Option<(Vec<String>, Vec<i32>, Vec<String>, Vec<String>, Vec<String>)>,
+    LNetState,
+) {
+    if lnet_data.net.len() == 0 {
```
Comment (utopiabound) [inline iml-services/iml-network/src/main.rs:168]: shouldn't this still be `lnet_data.net.is_empty()`?

File: iml-services/iml-network/src/main.rs [hunk 3]
```diff
@@ -175,57 +206,242 @@ fn parse_lnet_data(
             },
         );
 
-    Some(xs)
+    if xs.0.len() > 0 {
```
Comment (utopiabound) [inline iml-services/iml-network/src/main.rs:209]: also an `is_empty()` check?

**Review by utopiabound — APPROVED** [review]
Some minor nits, but looks good

**Pull request closed without merging**